### ISSUE TITLE
Wrong `JsonStreamContext` in `DeserializationProblemHandler` when reading `TokenBuffer` content

### ISSUE THREAD

**Issue by pgunia** [FasterXML/jackson-databind#1440]
I'm using a DeserializationProblemHandler to record unknown properties during the deserialisation of my entity instances. I use the following code to retrieve the full path through the JSON object. 

```
public boolean handleUnknownProperty(final DeserializationContext ctxt, final JsonParser p, final JsonDeserializer<?> deserializer,
        final Object beanOrClass, final String propertyName)
        throws IOException {

        final JsonStreamContext parsingContext = ctxt.getParser().getParsingContext();
        final List<String> pathList = new ArrayList<>();
        pathList.add(propertyName);

        addParent(parsingContext, pathList);
        Collections.reverse(pathList);
        final String path = String.join(".", pathList);

        deserializationProblemLog.addUnknownProperty(path);
        return super.handleUnknownProperty(ctxt, p, deserializer, beanOrClass, propertyName);

    }

    public DeserializationProblem getDeserializationProblemLog() {
        return deserializationProblemLog;
    }

    private void addParent(final JsonStreamContext streamContext, final List<String> pathList) {
        if (streamContext != null && streamContext.getCurrentName() != null) {
            pathList.add(streamContext.getCurrentName());
            addParent(streamContext.getParent(), pathList);
        }
    }
```

In general, this code works, but depending on the input JSON the recursion fails from time to time. For the following input, the result is as expected.
`"object": {
	"invalid_a": "value_a",
	"id": "ID"
}`
Output:
`object.invalid_a`

If I now add another line to my inout, namely 

`"object": {
	"type":"some_type",
	"invalid_a": "value_a",
	"id": "ID",
	"invalid_e": "null"
}`

the JsonStreamContext is no longer correct and my recursion breaks. I get the following result:

```
object.invalid_e.invalid_a",
object.invalid_e.invalid_e"
```
It seems to be connected to the "null" value of the last property. Nevertheless, it is not always reproducible. I'm using the standard ObjectMapper, only set `DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)`

I'm using version 2.8.4 of jackson-core, jackson-databind and jackson-annotations.

**Comment by cowtowncoder** [comment]
Would it be possible to get POJOs used when the problem occurs? I suspect discrepancy could be due to buffering (when using `@JsonCreator`, or polymorphic types, for example).

**Comment by pgunia** [comment]
Here are the entities that I use.  Both contain @JsonCreator annotated builder methods. 

```
@JsonSerialize(include = Inclusion.NON_NULL)
public class Activity implements Serializable {

    private ActivityEntity actors;

    private String verb;

    private ActivityEntity object;

    @Nullable
    private ActivityEntity target;

    private Activity(final Builder builder) {
        this.actor = builder.actor;
        this.verb = builder.verb;
        this.object = builder.object;
        this.target = builder.target;


    @JsonCreator
    public static Activity activityBuilder(@JsonProperty("actor") final ActivityEntity actor,
        @JsonProperty("verb") final String verb,
        @JsonProperty("object") final ActivityEntity object,
        @JsonProperty("target") final ActivityEntity target,
    ) {
        return new Builder(actor, verb, object)
            .setTarget(target)
            .build();
    }

    public static class Builder {

        private final ActivityEntity actor;
        private final String verb;
        private final ActivityEntity object;
        private ActivityEntity target;

        public Builder(final Activity activity) {
            this.actor = activity.actor;
            this.verb = activity.verb;
            this.object = activity.object;
            this.target = activity.target;
        }

        public Builder(final ActivityEntity actor, final String verb, final ActivityEntity object) {
            this.actor = actor;
            this.verb = verb;
            this.object = object;
        }

        public static Builder getBuilder(final ActivityEntity actor, final String verb, final ActivityEntity object) {
            return new Builder(actor, verb, object);
        }

        public Builder setTarget(final ActivityEntity target) {
            this.target = target;
            return this;
        }

        public Activity build() {
            return new Activity(this);
        }
    }
}

```

and 


```
@JsonSerialize(include = Inclusion.NON_NULL)
public class ActivityEntity implements Serializable {

    @Nullable
    private String id;

    @Nullable
    private String type;

    @Nullable
    private String status;

    @Nullable
    private String context;

    @JsonCreator
    public ActivityEntity(@JsonProperty("id") final String id, @JsonProperty("type") final String type) {
        this.id = id;
        this.type = type;
    }
}
```

**Comment by cowtowncoder** [comment]
Ok the problem is probably related to buffering needed by `@JsonCreator`, and linkage of parent context from buffer (which only holds a small set of tokens) to original parser (which has full parent stack). Worse, tokens are first added in buffer, but only replayed later on, with possibly different parent context -- that is, original parser may have been advanced to have different parent stack than when tokens were originally read.

I think what would be best here would be a small (as small as possible) reproduction; I can trouble-shoot parent linkage and see if problem can be fixed, or at least handling improved.

**Comment by pgunia** [comment]
Is there something I can do to help you with that?

**Comment by cowtowncoder** [comment]
@pgunia Just a reproduction, unit test I could run to show it happening, would be big help. Of course anything beyond that wrt debugging where discrepancy happens would help too.

**Comment by pgunia** [comment]
@cowtowncoder I've written some code that reproduces the issue. Take a look at the unit tests in this repo: 
https://github.com/pgunia/testStreamContext

**Comment by cowtowncoder** [comment]
@pgunia Ok. How do I build it? (seems to use Gradle, but I'm not too familiar with its use)

**Comment by pgunia** [comment]
@cowtowncoder Yep, it's a gradle project. You should be able to import it into your preferred IDE directly via the GitHub Link. The IDE should take care of the rest. At least Eclipse and IntelliJ come with integrated Gradle support. The IDE will then download the required dependencies (only Jackson, take a look at build.gradle) and you should then simply run the Unit tests to see the behaviour. Let me know if you need additional info on this.

**Comment by cowtowncoder** [comment]
ok, I can do this. I just vaguely recall there being command-line tool (`gradlew`?) or a way to bootstrap from cli. But IDE is fine too.

**Comment by cowtowncoder** [comment]
One minor thing on test: instead of using `context.getParser()`, `handleUnknownProperty()` should use parser it is being handed. Context does not necessarily have the currently active one during handling of buffered entries (it only has the original, "main" parser).
Changing this does not eliminate the problem currently but does make behavior bit more consistent.

**Comment by cowtowncoder** [comment]
Note to myself: `TokenBuffer` needs to link context from `parser` it is being given for construction (if any).
But it may also be necessary to peel off one level, or otherwise adjust, since location is likely to be slightly off if simply using whatever exists. Plus there's the slight risk of messing up parser's context if not carefully checking that parent stack matches.

**Comment by cowtowncoder** [comment]
@pgunia Ok this took bit longer and was somewhat hairy to fix but I _think_ I got it resolved for `master`. Fix didn't make it in `2.9.0.rc3` but should be available via local build. I would appreciate independent verification (and reports if it's failing in some case(s)); but provided test passes, assuming I did not somehow mess it. Code is now changed so that contextual path from `TokenBuffer` goes back to (a copy of) original context, so it should retain correct information regarding logical position (nesting, current name), although not full location information (but only starting location of the first buffered token).
